### ISSUE TITLE
TTY Output Regression

### ISSUE THREAD

**Issue by tianon** [moby/moby#33101]
**Description**

When using a recent `master` build including https://github.com/moby/moby/pull/33007 (to help kick the tires on the updated console-handling code), TTY output comes out garbled (presumably due to the lack of `\r`):
```console
$ ./docker -H tcp://docker:2375 run -it --rm bash
bash-4.4# echo hi
                 hi
                   bash-4.4# 
```

**Steps to reproduce the issue:**
1. create a master build (builds from https://master.dockerproject.org appear to be broken since https://github.com/moby/moby/pull/32694 was merged)
2. launch `dockerd`
3. run a container with an interactive shell and then invoke a command or two

**Describe the results you received:**

Cascading output!

**Describe the results you expected:**

Output aligned on the left margin, as in previous releases.

**Output of `docker version`:**

```console
$ ./docker -H tcp://docker:2375 version
Client:
 Version:      library-import
 API version:  1.30
 Go version:   go1.8.1
 Git commit:   library-import
 Built:        library-import
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-dev
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.1
 Git commit:   4b846a1
 Built:        Tue May  9 05:00:50 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```console
$ ./docker -H tcp://docker:2375 info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.06.0-dev
Storage Driver: vfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: d24f39e203aa6be4944f06dd0fe38a618a36c764
runc version: 992a5be178a62e026f4069f443c6164912adbf09
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.63-gentoo
Operating System: Alpine Linux v3.5 (containerized)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.4GiB
Name: 33784326e3d3
ID: QKJH:XFQL:UCOR:56QQ:RZES:4YEQ:625G:3RQF:BAJJ:3TUW:Z4IE:GKE6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: tianon
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I'm running the daemon via Docker-in-Docker, but my client is running directly on the host (connecting via TCP).  I also tried running my host's Docker client (17.04.0-ce) against the master daemon and got the same results.

**Edit:** I've reproduced on both `4.4.63-gentoo` and `4.9.0-2-amd64` (from Debian).

cc @crosbymichael @cyphar

**Comment by tianon** [comment]
My daemon is currently running with `vfs`, but I reproduced the same behavior with `overlay2` on a different host (same binaries).

I'm running my daemon container with `--privileged` and a volume on `/var/lib/docker`.

**Comment by tianon** [comment]
Here's another couple examples (showing that it's not related to `getline`):
```console
$ ./docker -H tcp://docker:2375 run -it --rm bash -c 'echo hi; echo hi2'
hi
  hi2
     $
```
```console
$ ./docker -H tcp://docker:2375 run -it --rm alpine:3.5 sh -c 'echo hi; echo hi2'
hi
  hi2
     $
```

**Comment by tianon** [comment]
(I've got some static binaries of 4b846a125b1618e19b389bcdf88b57bef0a05817 built via https://github.com/tianon/dockerfiles/blob/59ac82fbdc08724570652907f75326b38ce90ec2/docker-master/Dockerfile.build up at https://gist.github.com/tianon/3b19f345aca7a804919f76fe45ccac06.)

**Comment by cyphar** [comment]
Grrr. This is related to the changes I made to runc which un-set the `ONLCR` termios flag -- which means that `runc` now acts like a normal program's output (the trailing `\r\n` was breaking a bunch of our tests and will cause countless issues in shell scripts).

However this appears to break Docker in certain cases (which I haven't been able to reproduce), and it looks to me like there's something weird going on when the bytes are being shifted from `runc` into `containerd` and then into Docker. To be clear the current semantics from `runc` make sense if you were a normal program (creating a pty inside runc results in the extra `\r` which shouldn't be seen IMO).

**Comment by tianon** [comment]
I first saw it while running Docker inside a systemd-nspawn container running Gentoo (connecting via `machinectl shell`), which is kind of a hacky environment, so I verified by building and running the pristine static binaries off master in a Docker container -- do you think it's related to being inside a container?  Would it be helpful for me to try reproducing on a VM?

**Comment by crosbymichael** [comment]
@tianon is your docker client and daemon synced up?  We added back `OPOST` to sync the two sides of the tty.  I'm guessing your client is not in sync

**Comment by crosbymichael** [comment]
this could be something with cgo and a static build

**Comment by tianon** [comment]
@crosbymichael I built by copying the `moby/moby` repo and running the `install-binaries.sh` script to get the extra binaries (like the `dockercli` binary), so they should be synced

The first place I saw the behavior was inside an nspawn container where I was using dynamic binaries and a unix socket (rather than TCP).

I've built an image from the binaries I posted in that gist, and pushed it to `tianon/docker-master:4b846a125b1618e19b389bcdf88b57bef0a817-dind` (client-focused in `tianon/docker-master:4b846a125b1618e19b389bcdf88b57bef0a817`), so that the reproducer is easier to show:
```console
$ docker run -dit --privileged --name docker tianon/docker-master:4b846a125b1618e19b389bcdf88b57bef0a05817-dind --storage-driver aufs
8376a2de420f0a16016df27cf65c294311832964be9c6d6e2243d4c038172c6a
$ docker logs --tail=1 docker
INFO[0001] API listen on /var/run/docker.sock           
$ docker exec -it docker docker run -it --rm bash
Unable to find image 'bash:latest' locally
latest: Pulling from library/bash
627beaf3eaaf: Pull complete 
45c9d6ecfd7f: Pull complete 
79225b1f8264: Pull complete 
Digest: sha256:6d376b5826328395ed0fd414165b68c324732547e2530a1b6bf15fac1952e1cc
Status: Downloaded newer image for bash:latest
bash-4.4# echo hi
                 hi
                   bash-4.4# echo noooo
                                       noooo
                                            bash-4.4# 
```

**Comment by crosbymichael** [comment]
@cyphar its sane terminal that is doing this even with us syncing up OPOST

**Comment by tianon** [comment]
So this is likely related to https://github.com/opencontainers/runc/pull/1272? (which describes the same symptoms as my Docker issue)

**Comment by crosbymichael** [comment]
idk, i thought I added that back to docker

**Comment by d** [comment]
I'm still seeing this issue in latest Docker CE edge. I'm attaching output of `docker info`:

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 3
Server Version: 17.06.0-ce-rc1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
runc version: 992a5be178a62e026f4069f443c6164912adbf09
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.30-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: moby
ID: 
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 18
 Goroutines: 30
 System Time: 2017-06-06T03:54:34.104657404Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Comment by thaJeztah** [comment]
@d thanks for reporting; we're aware of this issue, and there's a separate issue tracking this for Docker for Mac, see https://github.com/moby/moby/issues/33489. It's resolved in "master" through https://github.com/moby/moby/pull/33502, which will be part of 17.06.0-ce-rc2

**Comment by g0t4** [comment]
@thaJeztah - any estimate on when 17.06.0-ce-rc2 will be released?

**Comment by thaJeztah** [comment]
@g0t4 rc2 is available;  the issue is a result of a breaking change in RunC upstream; initially, we decided to patch docker to accommodate for that change (this fix is in rc2) but that doesn't solve the problem if the client version doesn't match the daemon version, or if a different client (such as Docker Compose) is used to communicate with the API.

We decided to fork RunC, and carry a patch that reverts the breaking change (see https://github.com/moby/moby/pull/33577) -  That fix will be in 17.06.0-rc3. The breaking change is still being discussed with the RunC maintainers upstream to see if it should be reverted altogether.

So rc2 fixes it for now, but there will be a follow up fix in rc3
